File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        map<int,int>m;
        stack<int>st;
        int n=nums2.size();
        for(int i=n-1;i>=0;i--){
            while(!st.empty()){
                if(st.top()>nums2[i]){
                m[nums2[i]]=st.top();
                break;
                }
                st.pop();
            }
            if(st.empty()){
                m[nums2[i]]=-1;
            }
           st.push(nums2[i]); 
        }
        vector<int> ans;
        for(int i=0;i<nums1.size();i++){
            ans.push_back(m[nums1[i]]);
        }

        return ans;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer,Integer>m=new HashMap();
        Stack<Integer>st=new Stack<>();

        for(int n:nums2){
            while(!st.isEmpty() && st.peek()<n){
                m.put(st.peek(),n);
                st.pop();
            }
            st.add(n);
        }

         while(!st.isEmpty()){
                m.put(st.peek(),-1);
                st.pop();
            }

            int[] res=new int[nums1.length];
            for(int i=0;i<nums1.length;i++){
                res[i]=m.get(nums1[i]);
            }

            return res;
    }
}